content/post/2018-04-02-sct-encoding.markdown

```diff
@@ -0,0 +1,439 @@
+      # SignatureAndHashAlgorithm (ecdsa-sha256)
+      0403
+      # opaque signature<0..2^16-1>;
+      # length 0x0046
+      0046
+        # DER-encoded Ecdsa-Sig-Value - RFC 4492
+        30 # SEQUENCE
+        44 # length 0x44 bytes
+          02 # r INTEGER
+          20 # length 0x20 bytes
+            # value
+            7E1FCD1E9A2BD2A50A0C81E713033A0762340DA8F91EF27A48B3817640159CD3
+          02 # s INTEGER
+          20 # length 0x20 bytes
+            # value
+            659FE9F1D880E2E8F6B325BE9F18956D17C6CA8A6F2B12CB0F55FB70F759A419
+      # opaque SerializedSCT<1..2^16-1>
+      # length 0x77 bytes
+      0077
+      # Version sct_version v1(0)
+      00
+      # LogID id (aka opaque key_id[32])
+      293C519654C83965BAAA50FC5807D4B76FBF587A2972DCA4C30CF4E54547F478
+      # uint64 timestamp (milliseconds since the epoch)
+      000001627313EB2A
+      # CtExtensions extensions (zero-length array)
+      0000
+      # digitally-signed struct - RFC 5426
+      # SignatureAndHashAlgorithm (ecdsa-sha256)
+      0403
+      # opaque signature<0..2^16-1>;
+      # length 0x0048
+      0048
+        # DER-encoded Ecdsa-Sig-Value - RFC 4492
+        30 # SEQUENCE
+        46 # length 0x46 bytes
+          02 # r INTEGER
+          21 # length 0x21 bytes
+            # value
+            00AB72F1E4D6223EF87FC68491C208D29D4D57EBF47588BB7544D32F9537E2CEC1
+          02 # s INTEGER
+          21 # length 0x21 bytes
+            # value
+            008AFFC40CC6C4E3B24578DADE4F815ECBCE2D57A579342119A1E65BC7E5E69CE2
+```
+
+One surprising thing you might notice: In the first SCT, r and s are twenty
+bytes long. In the second SCT, they are both twenty-one bytes long, and have a
+leading zero. Integers in DER are two's complement, so if the leftmost bit is
+set, they are interpreted as negative. Since r and s are positive, if the
+leftmost bit would be a 1, an extra byte has to be added so that the leftmost
+bit can be 0.
+
+This is a little taste of what goes into encoding a certificate. I hope it was
+informative! If you'd like to learn more, I recommend "[A Layman's Guide to a
+Subset of ASN.1, BER, and DER](http://luca.ntop.org/Teaching/Appunti/asn1.html)."
+
+<a name="poison"></a>Footnote 1: A "poison extension" is defined by [RFC 6962
+section 3.1](https://tools.ietf.org/html/rfc6962#section-3.1):
+
+```
+The Precertificate is constructed from the certificate to be issued by adding a special
+critical poison extension (OID 1.3.6.1.4.1.11129.2.4.3, whose
+extnValue OCTET STRING contains ASN.1 NULL data (0x05 0x00))
+```
+
+In other words, it's an empty extension whose only purpose is to ensure that
+certificate processors will not accept precertificates as valid certificates. The
+specification ensures this by setting the "critical" bit on the extension, which
+ensures that code that doesn't recognize the extension will reject the whole
+certificate. And code that does recognize the extension specifically as poison
```

Comment:
nit: leading this last sentence with "And" is clunky. Can you join it with the prev?